File: 0-1-Knapsack/0-1-Knapsack.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int knapsack(int, int *, int *, int);
void printItems(int **, int *, int *, int, int);

int main()
{
    int n, W;
    cout << "Enter the number of items: ";
    cin >> n;
    cout << "Enter the capacity of the knapsack: ";
    cin >> W;
    cout << "Enter the weights of the items: ";
    int *weights = new int[n];
    for (int i = 0; i < n; i++)
    {
        cin >> weights[i];
    }
    cout << "Enter the profits of the items: ";
    int *profits = new int[n];
    for (int i = 0; i < n; i++)
    {
        cin >> profits[i];
    }
    int maxProfit = knapsack(W, weights, profits, n);
    cout << "The maximum profit is: " << maxProfit << endl;
    delete[] weights;
    delete[] profits;
    return 0;
}

int knapsack(int W, int *weights, int *profits, int n)
{
    int **c = new int *[n + 1];
    for (int i = 0; i <= n; i++)
    {
        c[i] = new int[W + 1];
    }

    for (int i = 0; i <= n; i++)
    {
        c[i][0] = 0;
    }

    for (int j = 0; j <= W; j++)
    {
        c[0][j] = 0;
    }

    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= W; j++)
        {
            int weight = weights[i - 1];
            if (weight <= j)
            {
                c[i][j] = std::max(c[i - 1][j], profits[i - 1] + c[i - 1][j - weight]);
            }
            else
            {
                c[i][j] = c[i - 1][j];
            }
        }
    }

    printItems(c, weights, profits, n, W);

    int result = c[n][W];

    for (int i = 0; i <= n; i++)
    {
        delete[] c[i];
    }
    return result;
}

void printItems(int **c, int *weights, int *profits, int n, int W)
{
    int i = n, j = W;
    cout << "Items selected : [";
    while (i > 0 && j > 0)
    {
        if (c[i][j] != c[i - 1][j])
        {
            cout << i;
            j = j - weights[i - 1];
            if (j > 0)
            {
                cout << ", ";
            }
        }
        i--;
    }
    cout << "]" << endl;
}

File: 0-1-Knapsack/0-1-Knapsack.py
def knapsack(W, weights, profits, n):
    c = [[0] * (W + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for j in range(1, W + 1):
            weight = weights[i - 1]
            if weight <= j:
                c[i][j] = max(c[i - 1][j], profits[i - 1] + c[i - 1][j - weight])
            else:
                c[i][j] = c[i - 1][j]

    print_items(c, weights, profits, n, W)

    return c[n][W]

def print_items(c, weights, profits, n, W):
    i, j = n, W
    result = []
    while i > 0 and j > 0:
        if c[i][j] != c[i - 1][j]:
            result.append(i)
            j -= weights[i - 1]
        i -= 1

    print("Items selected: [{}]".format(", ".join(map(str, result[::-1]))))

if __name__ == "__main__":
    n = int(input("Enter the number of items: "))
    W = int(input("Enter the capacity of the knapsack: "))
    
    weights = list(map(int, input("Enter the weights of the items: ").split()))
    profits = list(map(int, input("Enter the profits of the items: ").split()))

    max_profit = knapsack(W, weights, profits, n)
    print("The maximum profit is:", max_profit)
